pypesto/sampling/theano.py

```diff
@@ -10,31 +12,90 @@
     tt = NullType = None
 
 
+class CachedObjective:
```

Comment:
Would prefer if this class was derived from `pypesto.ObjectiveBase`

Reply:
this may be overly complicated here, as we effectively only require `__call__` with sensi orders 0 and 1, and only in this one place. If there is more general use for this class (it effectively reimplements a little bit of history functionality) one may consider refactoring.

Reply:
I don't see how this would be more complicated than what is currently done. Just have a look at `https://github.com/ICB-DCM/pyPESTO/blob/develop/pypesto/objective/aggregated.py`. Instead of piping through `__call__` one would just pipe through `call_unprocessed`. The sensi_order check could be directly implemented in check_sensi_orders/check_mode, which replaces the need to reimplement `has_grad`.
Additional benefits are that it becomes easier to reuse this in other places (we may think about also using this for optimization) and ensures that changes to the objective base class are properly propagated to this class.

Reply:
I also thought about implementing the objective interface, but being used just here (at the moment) I decided to keep it simple. It can be certainly be done. It is not my first priority, but after all the other pieces regarding pymc3 are working I'll try to do it.

Reply:
Now `CachedObjective` subclasses `ObjectiveBase`. It could still be generalized (at the moment it only supports `mode=MODE_FUN` and no hessians), but I would leave it for future improvements.